### PR TITLE
Arity check for shift and added some warnings

### PR DESCRIPTION
**Pull request by tscrim**
### 📚 Description

This address the arity problem noted in #35261. It also removes a `TODO` in a `revert()` implementation by making the behavior the same across different implementations. Warnings are added for the user for the assumptions made by the code.

### 📝 Checklist

- [x] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [x] I have linked an issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### ⌛ Dependencies

- #35127 Avoiding a long test.
- #35254 Avoiding merge conflicts.
- #35291 Making sure these changes work together.
- #35265 For some slight simplification of the new stream's `__getitem__`.

### CODE REVIEW

File: src/sage/rings/lazy_series.py [hunk 1]
```diff
@@ -657,6 +657,22 @@ def shift(self, n):
             sage: f.shift(3)
             1/(5^t) + 2/6^t
 
+        .. WARNING::
+
+            When working with an inexact power series (e.g., defined
```
Comment (mantepse) [inline src/sage/rings/lazy_series.py]: I strongly prefer "working with a power series which is not known to be eventually constant" here and elsewhere, to avoid confusion with series over inexact rings.

File: src/sage/rings/lazy_series.py [hunk 2]
```diff
@@ -657,6 +657,22 @@ def shift(self, n):
...
+            by a function) that has not computed its order and shifting
+            by too much can result in a Laurent series that does not
+            know it is a Laurent series::
```
Comment (mantepse) [inline src/sage/rings/lazy_series.py]: There is a grammatical / phrasing issue here, which I cannot quite pin down ("When ... shifting by too much can result ...")

More importantly, however, we should make it very clear that `fs` in the example below is broken, and we would actually prefer if we could raise an error, but we can't without breaking implicit definitions:
```
sage: L.<z> = LazyPowerSeriesRing(QQ)
sage: f = L(lambda n: n)
sage: g = f.shift(-2); g
1/z + 2 + 3*z + 4*z^2 + 5*z^3 + 6*z^4 + O(z^5)
sage: g.inverse()*g
1/z - 1 + 2*z - 3*z^2 + 5*z^3 - 8*z^4 + O(z^5)
```

Actually, there are options we might not have considered yet:
1. *always* return an element of the fraction field
2. return an element of the fraction field whenever `n` is negative
3. return an element of the fraction field, except if we *know* that the result is in `P` (i.e., `_approximate_order` is large enough.

Additionally / alternatively, we could 
1. have an optional keyword argument `check` (similar to what we do in `__call__`) which checks that  the valuation is large enough (i.e., computes enough terms to make sure that the result is in the parent
2. have an optional keyword argument `force` which implies that the expected parent is `P`, no matter what
3. have an optional keyword argument `parent` with the obvious meaning.

**Review by mantepse — COMMENTED** [review]
see inline comments.  I am not yet sure about the correct github workflow, sorry if I messed up.  Possibly the long inline comment should go into the comments to the pull request?

Oh, it does so automagically.

File: src/sage/rings/lazy_series.py (hunk 1 above)
Comment (tscrim) [reply src/sage/rings/lazy_series.py]: I think it is a bit of a mouthful, but that is a good point about confusion with inexact rings. I can't think of anything better at this point.

File: src/sage/rings/lazy_series.py (hunk 2 above)
Comment (tscrim) [reply src/sage/rings/lazy_series.py]: The sentence is slightly hard to parse, but it is correct. The key point is the "and" gives 2 necessary-but-not-sufficient conditions.

I would not say it is broken, nor would I say it should return an error. The behavior is well-defined, but the parent (and subsequently the element) just doesn't know enough to be fully correct. Strictly speaking, it is kind of a garbage-in-garbage-out scenario in my mind.

I am against options 1 and 2; while things would be more consistent (and hence, more manageable in some ways), it would also lead to very subtle surprises. 3 is certainly reasonable, although not without some subtleties. Alt 1 is somewhat reasonable, but it will put more burden on users with undefined series. Alt 3 is too cumbersome to allow fraction fields. Alt 2 is certainly a reasonable thing.

Actually, the more I think about it, the most consistent thing would be to just kill off any terms below the minimal valuation by default. Then taking a slight twist on Alt 2, we add a keyword that always returns an element in the fraction field.[^1] The current design is definitely an ease-of-implementation, but perhaps it was a mistake (likely mine). Changing it to this would make it the most consistent with almost no surprises. If we went this route, one remaining question is would we want to include a deprecation or not. 

[^1]: Strictly speaking, the Laurent field that contains the fraction field when working over a ring and not a field...

Comment (mantepse) [reply src/sage/rings/lazy_series.py]: (I don't understand the bit about `g.inverse()*g != 1` not being a bug, but I think that's not important)

I am all for consistent behaviour, it appears to me that inconsistencies which are convenient eventually become a pain.

Since Stream_shift shares the cache with its input stream, we will need a condition in `Stream_shift.__getitem__`.  I hope that this doesn't come with a performance penalty.

Also, `Stream_shift.__hash__` has to be adapted: `f.shift(-2)` and `f.shift(-1)` would then be different for `f = L(lambda n: 0 if n else 1)`.

Comment (tscrim) [reply src/sage/rings/lazy_series.py]: I reworked the shifting and the error message. There is a slight inconsistency between the known and not-known to be constant series, but this would be very complicated to make consistent I think. It is also a fairly fringe thing IMO.

There's a bit of a choice between inconsistent (but documented) behavior and simplicity of code in this case. It isn't really a pain to work with as I doubt anyone has encountered this situation without it being an actual bug in their code. For that reason, I don't think we need a deprecation warning, but I did add a flag just in case there is some user that was using the old behavior (now they have to always have the result as a Laurent series though).

The hash isn't a problem. Non-equal things can have the same hash. (There is a _very_ slight performance penalty when put into hast table objects such as a `dict`, but hash collisions are a general fact of life.) There are many more things that are an actual problem, where they are equal as FPS but have different hashes due to different constructions. However, solving that is likely a near-impossible task and definitely outside of the scope of this ticket.

**Comment by tscrim** [comment]
I made some additional changes to `revert` as well. This might need a dependency on that PR, but I will take care of that later.

File: src/sage/rings/lazy_series.py (hunk 2 above)
Comment (mantepse) [reply src/sage/rings/lazy_series.py]: Concerning the hash, I was writing about equal things that would have different hash.

For `f = L(lambda n: 0 if n else 1)`, the series `f.shift(-2)` and `f.shift(-1)` are equal if we do not push them into the fraction field, but will have the same hash.

Comment (tscrim) [reply src/sage/rings/lazy_series.py]: Yes, which is why I am a bit confused, equal implies the same hash is what we want.

Comment (mantepse) [reply src/sage/rings/lazy_series.py]: They would have differen hash, no?
```
        return hash((type(self), self._series))
```
Ah, no, they would actually have the same hash.  We should maybe make a note in `Stream_shift` that `__hash__` must not include the shift.

Comment (mantepse) [reply src/sage/rings/lazy_series.py]: I don't like the following:
```
sage: L.<x> = LazyPowerSeriesRing(QQ)
sage: f1 = 1/(1-x)
sage: f2 = L(lambda n: 1)
sage: f1.shift(-1)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In [17], line 1
----> 1 f1.shift(-Integer(1))

File ~/sage-trac/src/sage/rings/lazy_series.py:762, in LazyModuleElement.shift(self, n, use_fraction_field)
    760         if not init_coeff and not self._coeff_stream._constant:
    761             return P.zero()
--> 762     coeff_stream = Stream_exact(init_coeff,
    763                                 constant=self._coeff_stream._constant,
    764                                 order=valuation, degree=degree)
    765 else:
    766     coeff_stream = Stream_shift(self._coeff_stream, n)

File ~/sage-trac/src/sage/data_structures/stream.py:649, in Stream_exact.__init__(self, initial_coefficients, constant, degree, order)
    647 else:
    648     self._degree = degree
--> 649 assert order + len(initial_coefficients) <= self._degree
    651 # we remove leading and trailing zeros from
    652 # initial_coefficients
    653 
   (...)
    656 # different from constant, because __eq__ below would become
    657 # complicated otherwise
    658 for i, v in enumerate(initial_coefficients):

AssertionError: 
sage: f2.shift(-1)
1 + x + x^2 + x^3 + x^4 + x^5 + x^6 + O(x^7)
```

Comment (mantepse) [reply src/sage/rings/lazy_series.py]: And I really do not like
```
sage: g = f2.shift(-1)
sage: g
1 + x + x^2 + x^3 + x^4 + x^5 + x^6 + O(x^7)
sage: g[-1]
1
```

Comment (tscrim) [reply src/sage/rings/lazy_series.py]: Yes, that is really bad. It also exposes a number of other issues with overshifting that I don't have an easy or even moderate solution for right now. Creating a truncated shift stream is not easy if we want to share the cache without explicitly testing for the existence of a cache. The old behavior is starting to look attractive, albeit at the expense of a badly behaving answer if the user overshifts and doesn't use the optional argument. Or we otherwise just don't share the cache when we (might) truncate with perhaps some Python magic to convert to a usual stream if we eventually find that it is okay.

**Comment by mantepse** [comment]
Here is another bug (on 10.0.beta4, i.e., without any patch):
```
sage: L.<x> = LazyPowerSeriesRing(QQ)
sage: f1 = 1/(1-x)
sage: f2 = L(lambda n: 1)
sage: f2.shift(-1) == f1
AttributeError: 'Stream_shift' object has no attribute '_is_sparse'
```

**Comment by mantepse** [comment]
How about the following - always truncate?  (of course, exact case needs also adapting then, this is just proof of concept)

I have no idea what the speed penalty is.
```diff
diff --git a/src/sage/data_structures/stream.py b/src/sage/data_structures/stream.py
index 943fbf8face..2d0cfb26ce3 100644
--- a/src/sage/data_structures/stream.py
+++ b/src/sage/data_structures/stream.py
@@ -2721,8 +2721,9 @@ class Stream_shift(Stream):
 
     - ``series`` -- a :class:`Stream`
     - ``shift`` -- an integer
+    - ``minimal_order`` -- an integer
     """
-    def __init__(self, series, shift):
+    def __init__(self, series, shift, minimal_order=None):
         """
         Initialize ``self``.
 
@@ -2736,6 +2737,7 @@ class Stream_shift(Stream):
         """
         self._series = series
         self._shift = shift
+        self._minimal_order = minimal_order
         super().__init__(series._true_order)
 
     @lazy_attribute
@@ -2754,7 +2756,10 @@ class Stream_shift(Stream):
             [2, 3, 4, 5, 0]
         """
         # this is the true order, if self._series._true_order
-        return self._series._approximate_order + self._shift
+        if self._minimal_order is None:
+            return self._series._approximate_order + self._shift
+        return max(self._series._approximate_order + self._shift,
+                   self._minimal_order)
 
     def order(self):
         r"""
@@ -2768,8 +2773,10 @@ class Stream_shift(Stream):
             sage: s.order()
             3
         """
-        return self._series.order() + self._shift
-
+        if self._minimal_order is None:
+            return self._series.order() + self._shift
+        return max(self._series.order() + self._shift,
+                   self._minimal_order)
 
     def __getitem__(self, n):
         """
@@ -2786,7 +2793,9 @@ class Stream_shift(Stream):
             sage: [M[i] for i in range(6)]
             [0, 0, 0, 1, 2, 3]
         """
-        return self._series[n-self._shift]
+        if self._minimal_order is None or n >= self._minimal_order:
+            return self._series[n - self._shift]
+        return 0
 
     def __hash__(self):
         """
@@ -2823,6 +2832,7 @@ class Stream_shift(Stream):
             sage: M2 == Stream_shift(F, 2)
             True
         """
+        # TODO: check minimal_order
         return (isinstance(other, type(self)) and self._shift == other._shift
                 and self._series == other._series)
 
diff --git a/src/sage/rings/lazy_series.py b/src/sage/rings/lazy_series.py
index a2ec1653726..fea44092331 100644
index a2ec1653726..fea44092331 100644
--- a/src/sage/rings/lazy_series.py
+++ b/src/sage/rings/lazy_series.py
@@ -674,6 +674,7 @@ class LazyModuleElement(Element):
             sage: f.parent()
             Lazy Laurent Series Ring in x over Rational Field
         """
+        P = self.parent()
         if isinstance(self._coeff_stream, Stream_zero):
             return self
         elif isinstance(self._coeff_stream, Stream_shift):
@@ -690,14 +691,15 @@ class LazyModuleElement(Element):
                                         constant=self._coeff_stream._constant,
                                         order=valuation, degree=degree)
         else:
-            coeff_stream = Stream_shift(self._coeff_stream, n)
-        P = self.parent()
+            coeff_stream = Stream_shift(self._coeff_stream, n,
+                                        minimal_order=P._minimal_valuation)
+
-        # If we shift it too much, then it needs to go into the fraction field
-        # FIXME? This is different than the polynomial rings, which truncates the terms
-        if (coeff_stream._true_order
-            and P._minimal_valuation is not None
-            and coeff_stream._approximate_order < P._minimal_valuation):
-            P = P.fraction_field()
         return P.element_class(P, coeff_stream)
 
     __lshift__ = shift
```

**Comment by mantepse** [comment]
Concerning the equality failure, I guess it is best to make the comparison with `Stream_shift` in `Stream_exact` a special case.  Additionally, it is probably a good idea to check whether `isinstance(other, Stream_inexact)` before accessing the cache.

**Comment by mantepse** [comment]
roughly as follows (still needs some polishing for equality in `Stream_exact` and `Stream_shift`).

```diff
diff --git a/src/sage/data_structures/stream.py b/src/sage/data_structures/stream.py
index 943fbf8face..8bc01bf9b9e 100644
--- a/src/sage/data_structures/stream.py
+++ b/src/sage/data_structures/stream.py
@@ -847,14 +847,15 @@ class Stream_exact(Stream):
                     or self._constant != other._constant)
         # if other is not exact, we can at least compare with the
         # elements in its cache
-        if other._is_sparse:
-            for i in other._cache:
-                if self[i] != other._cache[i]:
-                    return True
-        else:
-            if other._offset > self._approximate_order:
-                return False
-            return any(self[i] != c for i, c in enumerate(other._cache, other._offset))
+        if isinstance(other, Stream_inexact):
+            if other._is_sparse:
+                for i in other._cache:
+                    if self[i] != other._cache[i]:
+                        return True
+            else:
+                if other._offset > self._approximate_order:
+                    return False
+                return any(self[i] != c for i, c in enumerate(other._cache, other._offset))
 
         return False
 
@@ -2095,7 +2096,7 @@ class Stream_plethysm(Stream_binary):
             sage: f = Stream_exact([1]) # irrelevant for this test
             sage: g = Stream_function(lambda n: s[n], True, 0)
             sage: h = Stream_plethysm(f, g, True, p)
-            sage: B = p[2, 2, 1](sum(s[i] for i in range(7)))
+            sage: B = p[2, 2, 1](sum(p(s[i]) for i in range(7)))
             sage: all(h.compute_product(k, Partition([2, 2, 1])) == B.restrict_degree(k) for k in range(7))
             True
         """
@@ -2721,8 +2722,9 @@ class Stream_shift(Stream):
 
     - ``series`` -- a :class:`Stream`
     - ``shift`` -- an integer
+    - ``minimal_order`` -- an integer
     """
-    def __init__(self, series, shift):
+    def __init__(self, series, shift, minimal_order=None):
         """
         Initialize ``self``.
 
@@ -2736,6 +2738,7 @@ class Stream_shift(Stream):
         """
         self._series = series
         self._shift = shift
+        self._minimal_order = minimal_order
         super().__init__(series._true_order)
 
     @lazy_attribute
@@ -2754,7 +2757,10 @@ class Stream_shift(Stream):
             [2, 3, 4, 5, 0]
         """
         # this is the true order, if self._series._true_order
-        return self._series._approximate_order + self._shift
+        if self._minimal_order is None:
+            return self._series._approximate_order + self._shift
+        return max(self._series._approximate_order + self._shift,
+                   self._minimal_order)
 
     def order(self):
         r"""
@@ -2768,8 +2774,10 @@ class Stream_shift(Stream):
             sage: s.order()
             3
         """
-        return self._series.order() + self._shift
-
+        if self._minimal_order is None:
+            return self._series.order() + self._shift
+        return max(self._series.order() + self._shift,
+                   self._minimal_order)
 
     def __getitem__(self, n):
         """
@@ -2786,7 +2794,9 @@ class Stream_shift(Stream):
             sage: [M[i] for i in range(6)]
             [0, 0, 0, 1, 2, 3]
         """
-        return self._series[n-self._shift]
+        if self._minimal_order is None or n >= self._minimal_order:
+            return self._series[n - self._shift]
+        return 0
 
     def __hash__(self):
         """
@@ -2823,6 +2833,7 @@ class Stream_shift(Stream):
             sage: M2 == Stream_shift(F, 2)
             True
         """
+        # TODO: check minimal_order
         return (isinstance(other, type(self)) and self._shift == other._shift
                 and self._series == other._series)
 
diff --git a/src/sage/rings/lazy_series.py b/src/sage/rings/lazy_series.py
index a2ec1653726..ec45c60daa8 100644
--- a/src/sage/rings/lazy_series.py
+++ b/src/sage/rings/lazy_series.py
@@ -670,10 +670,20 @@ class LazyModuleElement(Element):
 
             sage: L.<x> = LazyPowerSeriesRing(QQ)
             sage: f = x.shift(-3); f
-            x^-2
-            sage: f.parent()
-            Lazy Laurent Series Ring in x over Rational Field
+            0
+
+            sage: L.<x> = LazyPowerSeriesRing(QQ)
+            sage: f = 1 / (1 - x)
+            sage: g = f.shift(-1)
+            sage: f == g
+            True
+            sage: g[-1]
+            0
+            sage: h = L(lambda n: 1)
+            sage: h.shift(-1) == f
+            True
         """
+        P = self.parent()
         if isinstance(self._coeff_stream, Stream_zero):
             return self
         elif isinstance(self._coeff_stream, Stream_shift):
@@ -683,21 +693,22 @@ class LazyModuleElement(Element):
             else:
                 coeff_stream = self._coeff_stream._series
         elif isinstance(self._coeff_stream, Stream_exact):
-            init_coeff = self._coeff_stream._initial_coefficients
+            initial_coefficients = self._coeff_stream._initial_coefficients
             degree = self._coeff_stream._degree + n
             valuation = self._coeff_stream._approximate_order + n
-            coeff_stream = Stream_exact(init_coeff,
+            if (P._minimal_valuation is not None
+                and valuation < P._minimal_valuation):
+                initial_coefficients = initial_coefficients[P._minimal_valuation - valuation:]
+                degree = max(degree, P._minimal_valuation)
+                valuation = P._minimal_valuation
+                if not initial_coefficients and not self._coeff_stream._constant:
+                    return P.zero()
+            coeff_stream = Stream_exact(initial_coefficients,
                                         constant=self._coeff_stream._constant,
                                         order=valuation, degree=degree)
         else:
-            coeff_stream = Stream_shift(self._coeff_stream, n)
-        P = self.parent()
-        # If we shift it too much, then it needs to go into the fraction field
-        # FIXME? This is different than the polynomial rings, which truncates the terms
-        if (coeff_stream._true_order
-            and P._minimal_valuation is not None
-            and coeff_stream._approximate_order < P._minimal_valuation):
-            P = P.fraction_field()
+            coeff_stream = Stream_shift(self._coeff_stream, n,
+                                        minimal_order=P._minimal_valuation)
         return P.element_class(P, coeff_stream)
 
     __lshift__ = shift
```

**Comment by mantepse** [comment]
It seems to me (all tests pass) that `Stream_inexact.order` can actually be moved to `Stream`.

I am less sure about `is_nonzero`.

But I think that these problems can be solved, and that the real question is: do you agree we should define `shift` as truncating?

**Comment by tscrim** [comment]
I am already trying to do that, but the solutions are definitely not so simple.

**Comment by mantepse** [comment]
I am away for a few hours now, unfortunately.  The only thing that remains is `is_nonzero`, right?  If not, please post doctests!

**Comment by tscrim** [comment]
No problem. That gave me some time to work through a number of the issues that I came across. So after trying a few things, the best solution was to simply have a new stream class for the truncated shifting that only takes `Stream_inexact` objects. In the actual series `shift()` method, it will only get to that point when the base object is an inexact series; in particular, not a `Stream_shift` class as it pulls the shifting from that (which remains a good optimization IMO). I was also able to fix issues with the `order()` and `is_nonzero()`. Unfortunately I had to do a fair bit of duplication for the sparse `order()` case, but 'tis life. I remain a bit skeptical of how useful it was to make this consistent, but that's moot since it is done now.

**Comment by mantepse** [comment]
I have not yet read all of your proposal, but one thing seems quite clear to me: we can (and should) get rid of `use_fraction_field` again.  If you want your result to be in the fraction field, then it is trivial to do it yourself:
```
sage: L.<z> = LazyPowerSeriesRing(QQ)
sage: f = 1/(1-2*z)
sage: M.<x> = LazyLaurentSeriesRing(QQ)
sage: M(f).shift(-3)
x^-3 + 2*x^-2 + 4*x^-1 + 8 + 16*x + 32*x^2 + 64*x^3 + O(x^4)
```

**Comment by mantepse** [comment]
Concerning the duplication in `order`, I think it would be better to base this on #35265.

**Comment by tscrim** [comment]
True, but do we want to have a deprecation period? I am still leaning towards not, in which case we simply remove the `use_fraction_field` argument.

The changes in #35265 wouldn't make much difference here. Although I am fine either way. I guess I can finish the review of #35265 and merge that in here. I guess it saves one line...? What would you want to do?

**Comment by mantepse** [comment]
I'd prefer to have fraction_field not entering the code base and have this ticket be based on #35265, but I don't mind.

**Comment by tscrim** [comment]
Done and done. Note that this is now a backwards incompatible change (for something that might not be considered a bug) without a deprecation.

**Comment by mantepse** [comment]
It seems wrong to me that `Stream_truncated` messes with `_approximate_order`.

**Comment by tscrim** [comment]
Why? It needs to set its own independently of the `_series`. It doesn't change the _`series` one, but it needs to access that within `is_nonzero()` because it is using `_series._cache`.

**Comment by mantepse** [comment]
Oh yes, thank you, that makes sense.  Maybe you could add a comment
```
# We share self._series._cache but not self._series._approximate order
# self._approximate_order cannot be updated by self._series.__getitem__
```
 in the `__init__` just before `self._approximate_order = minimal_valuation`.

**Comment by tscrim** [comment]
Done, although I added it before `_cache` as I missed your recommendation for the location before I pushed.

I also noticed that we no longer need `_minimal_valuation`.

**Comment by mantepse** [comment]
There is one thing I'm not completely sure about yet: it seems that `Stream_truncated` will never set `self._true_order` in the dense case, and will generally not always use the best available bound for `_approximate_order`.

I think this is no problem.  Did you check / think about it?

**Comment by tscrim** [comment]
Good catch. Indeed, it doesn't. I have fixed this. I also made it go through the cache on initialization to see if it knows its true order or not as well. While it might have some slowdowns, I doubt it will be much and could end up speeding up some code due to avoiding redundant checks.

**Comment by mantepse** [comment]
Thank you!  Here are some polishing suggestions, nothing substantial - mainly avoiding to copy the cache.

```diff
diff --git a/src/sage/data_structures/stream.py b/src/sage/data_structures/stream.py
index 02a7449ce89..24e4a695986 100644
--- a/src/sage/data_structures/stream.py
+++ b/src/sage/data_structures/stream.py
@@ -2903,11 +2903,9 @@ class Stream_truncated(Stream_inexact):
         # self._approximate_order cannot be updated by self._series.__getitem__
         self._cache = series._cache
         self._shift = shift
-        ao = minimal_valuation
         # Try to find the true order based on the values already computed
-        true_order = False
         if self._is_sparse:
-            ao -= shift
+            ao = minimal_valuation - shift
             while ao in self._cache:
                 if self._cache[ao]:
                     self._true_order = True
@@ -2915,12 +2913,15 @@ class Stream_truncated(Stream_inexact):
                 ao += 1
             ao += shift
         else:
-            start = ao - (series._approximate_order + shift)
-            for val in self._cache[start:]:
-                if val:
-                    self._true_order = True
-                    break
-                ao += 1
+            ao = minimal_valuation - shift - series._approximate_order
+            try:
+                while not self._cache[ao]:
+                    ao += 1
+                self._true_order = True
+            except IndexError:
+                pass
+            ao += shift + series._approximate_order
+
         self._approximate_order = ao
 
     def __getitem__(self, n):
@@ -2948,7 +2949,7 @@ class Stream_truncated(Stream_inexact):
         """
         if n < self._approximate_order:
             return ZZ.zero()
-        ret = self._series[n-self._shift]
+        ret = self._series[n - self._shift]
         if not self._true_order:
             if self._is_sparse:
                 ao = self._approximate_order - self._shift
@@ -2960,12 +2961,14 @@ class Stream_truncated(Stream_inexact):
                 self._approximate_order = ao + self._shift
             else:  # dense case
                 offset = self._series._approximate_order + self._shift
-                start = self._approximate_order - offset
-                for val in self._cache[start:]:
-                    if val:
-                        self._true_order = True
-                        break
-                    self._approximate_order += 1
+                ao = self._approximate_order - offset
+                try:
+                    while not self._cache[ao]:
+                        ao += 1
+                    self._true_order = True
+                except IndexError:
+                    pass
+                self._approximate_order = ao + offset
         return ret
 
     def __hash__(self):
@@ -3043,7 +3046,7 @@ class Stream_truncated(Stream_inexact):
             cache = self._series._cache
             while True:
                 if n - self._shift in cache:
-                    if cache[n-self._shift]:
+                    if cache[n - self._shift]:
                         self._approximate_order = n
                         self._true_order = True
                         return n
@@ -3089,7 +3092,8 @@ class Stream_truncated(Stream_inexact):
             True
         """
         if self._is_sparse:
-            return any(c for n, c in self._series._cache.items() if n + self._shift >= self._approximate_order)
+            return any(c for n, c in self._series._cache.items()
+                       if n + self._shift >= self._approximate_order)
         offset = self._series._approximate_order + self._shift
         start = self._approximate_order - offset
         return any(self._cache[start:])
```

**Comment by tscrim** [comment]
It is much slower in Python to directly access the list at each point:
```
sage: L = list(range(100))
sage: %timeit sum(L[i] for i in range(5,100))
5.41 µs ± 353 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
sage: %timeit sum(L[5:])
1.17 µs ± 63.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
Thus, I want to avoid that. I don't think there will be too many problems with the slices since it just does a reference copy
```python
sage: L = []
sage: M = [1, 2, L]
sage: Ms = M[2:]
sage: Ms
[[]]
sage: L.append(5)
sage: M
[1, 2, [5]]
sage: Ms
[[5]]
```
and list slices are highly optimized in Python.  The cache would have to be **huge** for this to matter in terms of memory usage, and I think there would be other issues that would require modifying the source to get specialized workarounds before that would matter.

I didn't do the changes like
```diff
-                    if cache[n-self._shift]:
+                    if cache[n - self._shift]:
```
since I think the former is easier to parse and it is PEP8 compliant (since `[]` is a higher precedence operator). Although if you would prefer this, I don't care that much.

I think it is better to define `ao` outside the `if`-`else` statement as a slight hint to the reader that it is a fixed variable name used outside of these branches. Perhaps this is my C++ background and wanting to define variables at the top as much as possible.

**Comment by mantepse** [comment]
Even islice is slower.  How is that possible?  It is cheaper to create a copy of a list and then travese it than to traverse it?

**Comment by tscrim** [comment]
This is optimized core Python code for manipulating the lists (and likely just implemented as an appropriate `memcpy`, which is _very_ fast), whereas `itertools` is implemented on top of that and the `islice` essentially creates an iterator and advances it to the correct point.

**Comment by mantepse** [comment]
I (think I) hear what you say, I am just puzzled that in python, apparently, it can be faster to do A and B than to do just B.  After doing the `memcpy`, python still has to iterate over the elements in the list.

**Comment by tscrim** [comment]
Well, it is what it is. Are there any (additional) changes that I should do at this point?

**Comment by mantepse** [comment]
I cannot find the "approve" button :-(

**Review by mantepse — APPROVED** [review]
Thank you!

**Comment by tscrim** [comment]
Thank you.

**Comment by tscrim** [comment]
Trivial rebase.

**Comment by fchapoton** [comment]
This has broken the linter by introducing empty EXAMPLES blocks (and also changed 340 files..)

**Comment by fchapoton** [comment]
maybe you did not intend to include this "key tableaux crystals" ?

for the fix see #35418

**Comment by mantepse** [comment]
This commit absolutely does not belong to this ticket. I must say that I do not understand the notification above with the commit - it says "release manager"

**Comment by mantepse** [comment]
Should it be removed again, and if so, would that be possible?

**Comment by fchapoton** [comment]
I have removed the file in #35418

**Comment by tscrim** [comment]
@fchapoton Thank you. Sorry about this.
